### PR TITLE
Upgrade to Micronaut 4.0.x

### PR DESCRIPTION
**Pull request by alvarosanchez**

### CODE REVIEW

**Comment by alvarosanchez** [comment]
@yawkat please move this forward if you can

**Review by graemerocher — APPROVED** [review]

File: oraclecloud-function-http/src/main/java/io/micronaut/oraclecloud/function/http/FnMultiValueMap.java [hunk 1]
```diff
@@ -39,14 +41,27 @@
...
     private final Map<String, List<String>> map;
 
     /**
      * Default constructor.
-     * @param map The target map. Never null
+     *
+     * @param map               The target map. Never null
      */
     public FnMultiValueMap(Map<String, List<String>> map) {
+        this(map, new DefaultMutableConversionService());
```
Comment (yawkat) [inline oraclecloud-function-http/src/main/java/io/micronaut/oraclecloud/function/http/FnMultiValueMap.java:53]: I think this should be

```suggestion
        this(map, ConversionService.SHARED);
```

File: oraclecloud-function-http/src/main/java/io/micronaut/oraclecloud/function/http/HttpFunction.java [hunk 2]
```diff
@@ -40,11 +42,14 @@
...
     private ServletHttpHandler<InputEvent, OutputEvent> httpHandler;
 
+    private final ConversionService conversionService;
+
     /**
      * Default constructor.
      */
     @ReflectiveAccess
     public HttpFunction() {
+        this.conversionService = new DefaultMutableConversionService();
```
Comment (yawkat) [inline oraclecloud-function-http/src/main/java/io/micronaut/oraclecloud/function/http/HttpFunction.java:52]: same here

File: buildSrc/src/main/groovy/io.micronaut.build.internal.oraclecloud-module.gradle [hunk 3]
```diff
@@ -1,11 +1,10 @@
 plugins {
-    id 'io.micronaut.build.internal.common'
-    id 'io.micronaut.build.internal.module'
     id 'io.micronaut.build.internal.oraclecloud-base'
+    id 'io.micronaut.build.internal.module'
 }
 
 dependencies {
-    api platform("com.oracle.oci.sdk:oci-java-sdk-bom:$ociVersion")
+    api platform("com.oracle.oci.sdk:oci-java-sdk-bom:$oci3Version")
```
Comment (yawkat) [inline buildSrc/src/main/groovy/io.micronaut.build.internal.oraclecloud-module.gradle:7]: I think it should be possible to just change ociVersion now, and remove the special oci3Version property. We don't need the oci sdk 2 anymore.

**Review by yawkat — CHANGES_REQUESTED** [review]

**Comment by alvarosanchez** [comment]
So I was under the impression that we had to move away from `ConversionService.SHARED` for Micronaut 4. I have seen this change in many other modules that upgraded to 4. Is it not the case?

**Comment by yawkat** [comment]
@alvarosanchez ConversionService.SHARED is deprecated because it does not include all converters. You need to get the conversion service from the app context to get all converters. Calling the constructor does not fix this.

**Comment by alvarosanchez** [comment]
> You need to get the conversion service from the app context to get all converters. Calling the constructor does not fix this.

I did that in all places where the application context was available (or could be injected).

There is only one place where we can't do this AFAIK, and is the default constructor of `io.micronaut.oraclecloud.function.http.HttpFunction`, since that will be invoked via reflection by `com.fnproject.fn.runtime.EntryPoint`.

One more thing: `ConversionService.SHARED = new DefaultMutableConversionService()`, so reverting to the former is a bit pointless.

**Comment by yawkat** [comment]
Yea this one can't be replaced by the context conversion service.

@dstepanov what is the right thing to do here? `ConversionService.SHARED` or a new mutable one?

**Comment by alvarosanchez** [comment]
@graemerocher @yawkat this is potentially ready, please review.

**Review by yawkat — APPROVED** [review]

**Pull request merged**